umap/views.py

```diff
@@ -937,34 +941,83 @@ class DataLayerUpdate(FormLessEditMixin, GZipMixin, UpdateView):
-        # Simple response with only metadatas (client should not reload all data
-        # on save)
-        response = simple_json_response(
-            **self.object.metadata(self.request.user, self.request)
-        )
-        response["Last-Modified"] = self.last_modified
-        return response
+    def has_been_modified_since(self, if_unmodified_since):
+        if if_unmodified_since and self.last_modified != if_unmodified_since:
+            return True
+        return False
 
-    def is_unmodified(self):
-        """Optimistic concurrency control."""
-        modified = True
-        if_unmodified = self.request.META.get("HTTP_IF_UNMODIFIED_SINCE")
-        if if_unmodified:
-            if self.last_modified != if_unmodified:
-                modified = False
-        return modified
+    def merge(self, if_unmodified_since):
+        """
+        Attempt to apply the incoming changes to the document the client was using, and
+        then merge it with the last document we have on storage.
+
+        Returns either None (if the merge failed) or the merged python GeoJSON object.
+        """
+
+        # Use If-Modified-Since to find the correct version in our storage.
+        for name in self.object.get_versions():
+            path = os.path.join(settings.MEDIA_ROOT, self.object.get_version_path(name))
+            if if_unmodified_since == self.compute_last_modified(path):
+                with open(path) as f:
+                    reference = json.loads(f.read())
+                break
+        else:
+            # If the document is not found, we can't merge.
+            return None
+
+        # New data received in the request.
+        entrant = json.loads(self.request.FILES["geojson"].read())
+
+        # Latest known version of the data.
+        with open(self.path) as f:
+            latest = json.loads(f.read())
+
+        try:
+            merged_features = merge_features(
+                reference["features"], latest["features"], entrant["features"]
+            )
+            latest["features"] = merged_features
+            return latest
+        except ConflictError:
+            return None
 
     def post(self, request, *args, **kwargs):
         self.object = self.get_object()
         if self.object.map.pk != int(self.kwargs["map_id"]):
             return HttpResponseForbidden()
+
         if not self.object.can_edit(user=self.request.user, request=self.request):
             return HttpResponseForbidden()
-        if not self.is_unmodified():
-            return HttpResponse(status=412)
-        return super(DataLayerUpdate, self).post(request, *args, **kwargs)
+
+        ius_header = self.request.META.get("HTTP_IF_UNMODIFIED_SINCE")
+
+        if self.has_been_modified_since(ius_header):
+            merged = self.merge(ius_header)
+            if not merged:
+                return HttpResponse(status=412)
+
+            # Replace the uploaded file by the merged version.
```

Comment:
I'm not a fan of this "let's overwrite what the client sent us", do you have a better idea?